app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/liveview/LiveviewIoThread.java

```diff
@@ -0,0 +1,220 @@
+import java.util.UUID;
+
+import nodomain.freeyourgadget.gadgetbridge.deviceevents.GBDeviceEvent;
+import nodomain.freeyourgadget.gadgetbridge.impl.GBDevice;
+import nodomain.freeyourgadget.gadgetbridge.service.serial.GBDeviceIoThread;
+import nodomain.freeyourgadget.gadgetbridge.service.serial.GBDeviceProtocol;
+import nodomain.freeyourgadget.gadgetbridge.util.GB;
+
+public class LiveviewIoThread extends GBDeviceIoThread {
+    private static final Logger LOG = LoggerFactory.getLogger(LiveviewIoThread.class);
+
+    private static final UUID SERIAL = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
+
+    private final LiveviewProtocol mLiveviewProtocol;
+    private final LiveviewSupport mLiveviewSupport;
+
+
+    private BluetoothAdapter mBtAdapter = null;
+    private BluetoothSocket mBtSocket = null;
+    private InputStream mInStream = null;
+    private OutputStream mOutStream = null;
+    private boolean mQuit = false;
+
+    @Override
+    public void quit() {
+        mQuit = true;
+        if (mBtSocket != null) {
+            try {
+                mBtSocket.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    private boolean mIsConnected = false;
+
+
+    public LiveviewIoThread(GBDevice gbDevice, Context context, GBDeviceProtocol lvProtocol, LiveviewSupport lvSupport, BluetoothAdapter lvBtAdapter) {
+        super(gbDevice, context);
+        mLiveviewProtocol = (LiveviewProtocol) lvProtocol;
+        mBtAdapter = lvBtAdapter;
+        mLiveviewSupport = lvSupport;
+    }
+
+    @Override
+    public synchronized void write(byte[] bytes) {
+        if (null == bytes)
+            return;
+        LOG.debug("writing:" + GB.hexdump(bytes, 0, bytes.length));
+        try {
+            mOutStream.write(bytes);
+            mOutStream.flush();
+        } catch (IOException e) {
+            LOG.error("Error writing.", e);
+        }
+    }
+
+    @Override
+    public void run() {
+        mIsConnected = connect();
+        if (!mIsConnected) {
+            setUpdateState(GBDevice.State.NOT_CONNECTED);
+            return;
+        }
+        mQuit = false;
+
+        while (!mQuit) {
+            LOG.info("Ready for a new message exchange.");
+
+            try {
+                GBDeviceEvent deviceEvents[] = mLiveviewProtocol.decodeResponse(parseIncoming());
+                if (deviceEvents == null) {
+                    LOG.info("unhandled message");
+                } else {
+                    for (GBDeviceEvent deviceEvent : deviceEvents) {
+                        if (deviceEvent == null) {
+                            continue;
+                        }
+                        mLiveviewSupport.evaluateGBDeviceEvent(deviceEvent);
+                    }
+                }
+
+            } catch (IOException e) {
+                if (e.getMessage() != null && e.getMessage().contains("socket closed")) { //FIXME: this does not feel right
```

Comment:
To fix this you should probably catch java.net.SocketException before catching IOException.